### DISCUSSION TITLE
button in text entry

### DISCUSSION THREAD

**Discussion by iacore** [david-vanderson/dvui#37] [General]
I found an interesting layout.

![image](https://github.com/david-vanderson/dvui/assets/74560659/0ca69af4-b7cf-4637-8395-4797a6119d3c)

Code:
I think this layout is because `defer entry.deinit()` is called last.

```zig
        {
            const entry = try dvui.textEntry(@src(), .{ .text = &doit_buffer, .scroll_vertical_bar = .hide, .scroll_horizontal_bar = .hide }, .{ .expand = .horizontal });
            defer entry.deinit();
            const text = doit_buffer[0..entry.len];
            if (text.len > 0 and text[text.len - 1] == '\n') {
                text[text.len - 1] = 0;
                try this.tryDo(text[0 .. text.len - 1], &doit_buffer, .getit);
            }
            const box = try dvui.box(@src(), .horizontal, .{ .gravity_x = 1 });
            defer box.deinit();
            if (try dvui.button(@src(), "Do It", .{})) {
                try this.tryDo(text, &doit_buffer, .doit);
            }
            if (try dvui.button(@src(), "Get It", .{})) {
                try this.tryDo(text, &doit_buffer, .getit);
            }
        }
```

**Comment by david-vanderson** [comment]
Whoops!  Putting widgets in the corners of a TextLayoutWidget is a supported feature, but I didn't realize it flowed through to a TextEntryWidget.  I should probably disable that.

**Reply by iacore** [reply]
maybe this is a feature, not a bug

**Comment by david-vanderson** [comment]
Let's try supporting this use case.  I cleaned up the TextEntryWidget api and fixed some bugs around corner widgets and sizing.

The functions textEntry() and textLayout() now have comments showing where to add corner widgets.

Let me know if you find a good use for it!
